### PR TITLE
IGNITE-17059 Thin 3.0: Implement batch SQL API for java thin client

### PR DESCRIPTION
**Pull request by xtern**
https://issues.apache.org/jira/browse/IGNITE-17059

* Added support for `SQL_EXEC_BATCH` operation on java thin side.
* `BatchedArguments` has been reworked to prevent possible issue with its serialization.

### CODE REVIEW

File: modules/api/src/main/java/org/apache/ignite/sql/BatchedArguments.java [hunk 1]
```diff
@@ -18,24 +18,25 @@
...
 
 /**
  * Arguments for batch query execution.
- *
- * <p>TODO: replace inheritance with delegation.
- * TODO: add arguments length validation.
- * TODO: add named arguments support.
  */
-public class BatchedArguments extends ArrayList<List<Object>> implements List<List<Object>> {
+public final class BatchedArguments implements Iterable<List<Object>> {
```
Comment (ptupitsyn) [inline modules/api/src/main/java/org/apache/ignite/sql/BatchedArguments.java:29]: We have size and positional access, let's implement `List` interface.

File: modules/api/src/main/java/org/apache/ignite/sql/BatchedArguments.java [hunk 2]
```diff
@@ -73,8 +72,132 @@ public static BatchedArguments of(List<List<Object>> args) {
...
+     * Appends arguments to a batch.
+     *
+     * @param argsList Arguments list.
+     * @return {@code this} for chaining.
+     */
+    public BatchedArguments add(List<Object> argsList) {
+        Objects.requireNonNull(argsList, "argsList");
+
+        if (argsList.isEmpty()) {
+            throwEmptyArgumentsException();
```
Comment (ptupitsyn) [inline modules/api/src/main/java/org/apache/ignite/sql/BatchedArguments.java:90]: The user can modify the list after this method call, do we handle this?

File: modules/client/src/main/java/org/apache/ignite/internal/client/sql/ClientSql.java [hunk 3]
```diff
@@ -264,15 +258,57 @@ public <T> CompletableFuture<AsyncResultSet<T>> executeAsync(
...
+
+            w.out().packString(statement.query());
+            w.out().packBatchedArgumentsAsBinaryTupleArray(batch);
+            w.out().packLong(ch.observableTimestamp());
+        };
+
+        PayloadReader<BatchResultInternal> payloadReader = r -> {
+            ClientMessageUnpacker unpacker = r.in();
+
+            unpacker.skipValues(4); // skipping values that are not currently in use.
```
Comment (ptupitsyn) [inline modules/client/src/main/java/org/apache/ignite/internal/client/sql/ClientSql.java:286]: Please list every skipped value name here.

File: modules/client/src/main/java/org/apache/ignite/internal/client/sql/ClientSql.java [hunk 4]
```diff
@@ -264,15 +258,57 @@ public <T> CompletableFuture<AsyncResultSet<T>> executeAsync(
...
+
+            long[] updateCounters = unpacker.unpackLongArray();
+
+            if (unpacker.tryUnpackNil()) {
+                unpacker.skipValues(2);
```
Comment (ptupitsyn) [inline modules/client/src/main/java/org/apache/ignite/internal/client/sql/ClientSql.java:291]: Please add a comment explaining what is skipped an why.

File: modules/client-handler/src/main/java/org/apache/ignite/client/handler/requests/sql/ClientSqlExecuteBatchRequest.java [hunk 5]
```diff
@@ -102,16 +103,18 @@ public static CompletableFuture<Void> process(
...
+            String errorMessage,
+            UUID traceId) {
         out.packNil(); // resourceId
 
         out.packBoolean(false); // has row set
         out.packBoolean(false); // has more pages
         out.packBoolean(false); // was applied
         out.packLongArray(affectedRows); // affected rows
-        out.packShort(errorCode); // error code
+        out.packInt(errorCode); // error code
```
Comment (ptupitsyn) [inline modules/client-handler/src/main/java/org/apache/ignite/client/handler/requests/sql/ClientSqlExecuteBatchRequest.java:115]: We have standardized error serialization in https://github.com/apache/ignite-3/blob/d3ab33124127dd18f36727148100d632fffbd02e/modules/client-handler/src/main/java/org/apache/ignite/client/handler/ClientInboundMessageHandler.java#L498 .

Let's remove error details from this response, and return failed future instead.

**Review by ptupitsyn — CHANGES_REQUESTED** [review]

**Review by AMashenkov — APPROVED** [review]

File: modules/api/src/main/java/org/apache/ignite/sql/BatchedArguments.java (hunk 1 above)
Comment (xtern) [reply modules/api/src/main/java/org/apache/ignite/sql/BatchedArguments.java:29]: I'm not sure about this. This class is a part of public API.
java.util.List contains ~30 methods, that are not required by this container.
I think it's better to narrow down the functionality of this class and extend it later if needed.

File: modules/api/src/main/java/org/apache/ignite/sql/BatchedArguments.java (hunk 2 above)
Comment (xtern) [reply modules/api/src/main/java/org/apache/ignite/sql/BatchedArguments.java]: Thanks, I believe I fixed this.

File: modules/client/src/main/java/org/apache/ignite/internal/client/sql/ClientSql.java (hunk 3 above)
Comment (xtern) [reply modules/client/src/main/java/org/apache/ignite/internal/client/sql/ClientSql.java]: Done.

File: modules/client/src/main/java/org/apache/ignite/internal/client/sql/ClientSql.java (hunk 4 above)
Comment (xtern) [reply modules/client/src/main/java/org/apache/ignite/internal/client/sql/ClientSql.java:297]: Done.

File: modules/client-handler/src/main/java/org/apache/ignite/client/handler/requests/sql/ClientSqlExecuteBatchRequest.java (hunk 5 above)
Comment (xtern) [reply modules/client-handler/src/main/java/org/apache/ignite/client/handler/requests/sql/ClientSqlExecuteBatchRequest.java:115]: For SqlBatchException this doesn't work very well right now.
This exception includes an "update counter" which is lost during serialization.
This counter can be used by the user to understand how many updates were completed successfully (we have several tests for this).

File: modules/api/src/main/java/org/apache/ignite/sql/BatchedArguments.java (hunk 1 above)
Comment (ptupitsyn) [reply modules/api/src/main/java/org/apache/ignite/sql/BatchedArguments.java:29]: Ok.

File: modules/client-handler/src/main/java/org/apache/ignite/client/handler/requests/sql/ClientSqlExecuteBatchRequest.java (hunk 5 above)
Comment (ptupitsyn) [reply modules/client-handler/src/main/java/org/apache/ignite/client/handler/requests/sql/ClientSqlExecuteBatchRequest.java:115]: Ok, let's handle this as a separate ticket: https://issues.apache.org/jira/browse/IGNITE-22569

**Pull request merged**